### ISSUE TITLE
No detailed explanation if java-file is failed to be parsed

### ISSUE THREAD

**Issue by n-semikin** [openrewrite/rewrite#3645]
## What version of OpenRewrite are you using?
I am using

- Maven/Gradle plugin v5.9.1
- rewrite-migrate-java v2.1.1

## How are you running OpenRewrite?
I am using the Maven plugin, and my project is a single module project.
```xml
      <plugin>
        <groupId>org.openrewrite.maven</groupId>
        <artifactId>rewrite-maven-plugin</artifactId>
        <version>5.9.1</version>
        <configuration>
          <activeRecipes>
            <recipe>com.test.Test</recipe>
          </activeRecipes>
        </configuration>
        <dependencies>
          <dependency>
            <groupId>org.openrewrite.recipe</groupId>
            <artifactId>rewrite-migrate-java</artifactId>
            <version>2.1.1</version>
          </dependency>
          <dependency>
            <groupId>org.openrewrite</groupId>
            <artifactId>rewrite-core</artifactId>
            <version>8.8.0</version>
          </dependency>
        </dependencies>
      </plugin>
```

## What is the smallest, simplest way to reproduce the problem?
```java
package org.example.parsing_error;

import java.io.Serializable;
import java.util.Comparator;
import java.util.TreeMap;

class SomeClass {

    private void createStrings() {
        new TreeMap<>((Comparator<Integer> & Serializable) Integer::compare);
    }

}
```
This expression `(Comparator<Integer> & Serializable)` is called `intersection type cast` (according to ChatGPT :) ). It casts to the type that is compatible with both `Serializable` and `Comparator` interface. I don't know why it is added in my project and it works without it. So I'm going to remove it from my project. But I think you are interested in parsing a valid java syntax without issues. 

`rewrite.yml` file is
```yaml
---
type: specs.openrewrite.org/v1beta/recipe
name: com.test.Test
recipeList:
  - org.openrewrite.java.ChangeType:
      oldFullyQualifiedTypeName: com.sun.javafx.scene.control.skin.TableHeaderRow
      newFullyQualifiedTypeName: javafx.scene.control.skin.TableHeaderRow
      ignoreDefinition: false
```

Command to run is
```shell
./mvnw -U org.openrewrite.maven:rewrite-maven-plugin:run
```

## What did you expect to see?
no errors in the output

## What did you see instead?
```log
20231025_22:21:12.090 [WARNING] There were problems parsing some source files, run with --errors to see full stack traces
20231025_22:21:12.090 [WARNING] There were problems parsing src/main/java/org/example/parsing_error/SomeClass.java
```

## What is the full stack trace of any errors you encountered?
Adding `-X` option to the Maven doesn't add more details on the error:
```
20231025_22:47:44.419 [INFO] Project [TestRewrite] Resolving Poms...
20231025_22:47:47.230 [DEBUG] Project [TestRewrite] Base directory : '/Users/semikin/work/smarttab/temp/TestRewrite'
20231025_22:47:47.230 [DEBUG] Project [TestRewrite]   Collected Maven POM : '/Users/semikin/work/smarttab/temp/TestRewrite/pom.xml'
20231025_22:47:47.231 [DEBUG] Project [TestRewrite]   Maven Source : '/Users/semikin/work/smarttab/temp/TestRewrite/pom.xml'
20231025_22:47:47.253 [INFO] Project [TestRewrite] Parsing source files
20231025_22:47:47.788 [DEBUG] Project [TestRewrite] Scanned 6 java source files in main scope.
20231025_22:47:47.788 [DEBUG] Project [TestRewrite] Scanned 0 kotlin source files in main scope.
20231025_22:47:47.788 [DEBUG] Project [TestRewrite] Scanned 0 resource files in main scope.
20231025_22:47:48.025 [DEBUG] Project [TestRewrite] Scanned 0 java source files in test scope.
20231025_22:47:48.025 [DEBUG] Project [TestRewrite] Scanned 0 kotlin source files in main scope.
20231025_22:47:48.025 [DEBUG] Project [TestRewrite] Scanned 0 resource files in test scope.
20231025_22:47:48.271 [DEBUG] Project [TestRewrite] Parsed 14 additional files found within the project.
20231025_22:47:48.758 [WARNING] There were problems parsing some source files
20231025_22:47:48.758 [WARNING] There were problems parsing src/main/java/org/example/parsing_error/SomeClass.java
20231025_22:47:50.413 [INFO] Running recipe(s)...
20231025_22:47:50.427 [DEBUG] Using SLF4J as the default logging framework
20231025_22:47:50.451 [INFO] ------------------------------------------------------------------------
20231025_22:47:50.451 [INFO] BUILD SUCCESS
```

So currently in my project I have several issues related to java-file parsing. And **it takes quite a lot of time to figure out what expression causes a parsing issue**. 
I cannot figure out where to put `--errors` options specified in the first log. Probably you can direct me to the proper documentation.

Also trying to find the way of getting more details I found a [ParseExceptionResult](https://docs.openrewrite.org/recipes/core/findparsefailures). But it doesn't provide any output to the console. Seems it just put some notes in LST.

**Comment by ammachado** [comment]
This issue also happens with `JavaTemplate`s, but so far I couldn't get a reproducer. The recipes I had issues so far were `CombineSemanticallyEqualCatchBlocks` and `RemoveTryCatchFailBlocks`

**Comment by n-semikin** [comment]
Hey @timtebeek!
You helped me a lot in other issues) Do you have idea on how to get more info about parsing error?

**Comment by knutwannheden** [comment]
@n-semikin Thanks for reporting. For the error with the parsing of intersection type casts I have created a new issue: https://github.com/openrewrite/rewrite/issues/3651. So let's discuss any questions related to that in there.

We will keep this issue open for the request related to printing the parsing errors and warnings.

**Comment by n-semikin** [comment]
Thank you @knutwannheden for the updated.

My bigger issue is that log doesn't provide any info on what caused the issue. It blocks me for quite a long time, because it requires multiple code commenting and rewrite running. 
Can please provide info on how to turn on detailed parser logs?

**Comment by n-semikin** [comment]
Nice idea! If I got you correctly I'm able to add a RewriteTest that will process a specific file in the project. And while its processing there should be detailed logs. Will try to achieve that.

**Comment by knutwannheden** [comment]
Yes, that is the gist of it. The test checks that the source prints back out exactly to the same text that was parsed. When this isn't the case the build plugin will report the warning you saw. The test will however also print the diff of what was parsed and what was printed. That will then help us track the error down and fix it in our parser.

Thanks for your understanding!

**Comment by knutwannheden** [comment]
@n-semikin Meanwhile the intersection type casts are now supported: #3652

**Comment by TomChatt2** [comment]
I came across this thread because I'm just starting to try openrewrite (5.11.0), and am baffled by the message "there were problems with some source files, run with --errors ...".  I read the thread above, but I'm not quite following the suggestion. Could you spell out more explicitly what a user is supposed to do when they encounter that message? How do I get more information about which files are causing parse errors, and what I should do to fix those files?

**Comment by knutwannheden** [comment]
@TomChatt2 Thanks for reaching out. The message regarding `--errors` is indeed misleading, which is why we've kept this issue open.

Please note that these parsing errors are typically bugs in our parsers, rather than errors in your sources (unless you know that the source contains errors on purpose).

Currently, the easiest way is if you could somehow send us the whole source file (could also be via email), so that we can analyze it and create an issue for it.

If it is a Java source which is listed as causing errors, then it is most likely a Unicode escape sequence in Javadoc or possibly also an annotation on an array type's dimension (as in `String @Nullable []`).
